File: fujsen/src/CoreOperators/BaseOperator.php
<?php
/**
 * Base Operator Class
 * 
 * Provides common functionality for all @ operators including
 * validation, error handling, and context management.
 * 
 * @package TuskPHP\CoreOperators
 * @version 2.0.0
 */

namespace TuskPHP\CoreOperators;

use TuskPHP\Core\OperatorPlugin;

/**
 * Base Operator Class
 * 
 * All operators should extend this class to get common functionality
 * like validation, error handling, and context management.
 */
abstract class BaseOperator implements OperatorPlugin
{
    protected array $defaultConfig = [];
    protected array $requiredFields = [];
    protected array $optionalFields = [];
    protected string $version = '1.0.0';
    protected bool $debugMode = false;
    
    /**
     * Get operator name
     */
    abstract public function getName(): string;
    
    /**
     * Get operator version
     */
    public function getVersion(): string
    {
        return $this->version;
    }
    
    /**
     * Get operator schema
     */
    public function getSchema(): array
    {
        return [
            'name' => $this->getName(),
            'version' => $this->getVersion(),
            'description' => $this->getDescription(),
            'required_fields' => $this->requiredFields,
            'optional_fields' => $this->optionalFields,
            'default_config' => $this->defaultConfig,
            'examples' => $this->getExamples(),
            'error_codes' => $this->getErrorCodes()
        ];
    }
    
    /**
     * Get operator description
     */
    protected function getDescription(): string
    {
        return 'Base operator description - override in subclasses';
    }
    
    /**
     * Get usage examples
     */
    protected function getExamples(): array
    {
        return [
            'basic' => "@{$this->getName()}(\"example\")",
            'advanced' => "@{$this->getName()}({param: \"value\"})"
        ];
    }
    
    /**
     * Get error codes
     */
    protected function getErrorCodes(): array
    {
        return [
            'INVALID_CONFIG' => 'Invalid configuration provided',
            'EXECUTION_FAILED' => 'Operator execution failed',
            'TIMEOUT' => 'Operation timed out',
            'CONNECTION_FAILED' => 'Connection to external service failed'
        ];
    }
    
    /**
     * Validate configuration
     */
    public function validate(array $config): array
    {
        $errors = [];
        $warnings = [];
        
        // Check required fields
        foreach ($this->requiredFields as $field) {
            if (!isset($config[$field])) {
                $errors[] = "Required field '{$field}' is missing";
            }
        }
        
        // Validate field types and values
        $validation = $this->validateFields($config);
        $errors = array_merge($errors, $validation['errors']);
        $warnings = array_merge($warnings, $validation['warnings']);
        
        // Custom validation
        $customValidation = $this->customValidate($config);
        $errors = array_merge($errors, $customValidation['errors']);
        $warnings = array_merge($warnings, $customValidation['warnings']);
        
        return [
            'valid' => empty($errors),
            'errors' => $errors,
            'warnings' => $warnings
        ];
    }
    
    /**
     * Validate individual fields
     */
    protected function validateFields(array $config): array
    {
        $errors = [];
        $warnings = [];
        
        foreach ($config as $field => $value) {
            $validation = $this->validateField($field, $value);
            if (!empty($validation['errors'])) {
                $errors = array_merge($errors, $validation['errors']);
            }
            if (!empty($validation['warnings'])) {
                $warnings = array_merge($warnings, $validation['warnings']);
            }
        }
        
        return ['errors' => $errors, 'warnings' => $warnings];
    }
    
    /**
     * Validate a single field
     */
    protected function validateField(string $field, mixed $value): array
    {
        $errors = [];
        $warnings = [];
        
        // Check if field is known
        if (!in_array($field, array_merge($this->requiredFields, $this->optionalFields))) {
            $warnings[] = "Unknown field '{$field}'";
        }
        
        // Type validation
        $typeValidation = $this->validateFieldType($field, $value);
        if (!empty($typeValidation['errors'])) {
            $errors = array_merge($errors, $typeValidation['errors']);
        }
        
        return ['errors' => $errors, 'warnings' => $warnings];
    }
    
    /**
     * Validate field type
     */
    protected function validateFieldType(string $field, mixed $value): array
    {
        $errors = [];
        
        // Override in subclasses for specific type validation
        return ['errors' => $errors];
    }
    
    /**
     * Custom validation logic
     */
    protected function customValidate(array $config): array
    {
        // Override in subclasses for custom validation
        return ['errors' => [], 'warnings' => []];
    }
    
    /**
     * Execute operator
     */
    public function execute(array $config, array $context): mixed
    {
        try {
            // Merge with default configuration
            $config = array_merge($this->defaultConfig, $config);
            
            // Pre-execution hooks
            $this->beforeExecute($config, $context);
            
            // Execute the operator
            $result = $this->executeOperator($config, $context);
            
            // Post-execution hooks
            $this->afterExecute($config, $context, $result);
            
            return $result;
            
        } catch (\Exception $e) {
            $this->handleError($e, $config, $context);
            throw $e;
        }
    }
    
    /**
     * Execute the actual operator logic
     */
    abstract protected function executeOperator(array $config, array $context): mixed;
    
    /**
     * Pre-execution hook
     */
    protected function beforeExecute(array $config, array $context): void
    {
        if ($this->debugMode) {
            error_log("Executing @{$this->getName()} with config: " . json_encode($config));
        }
    }
    
    /**
     * Post-execution hook
     */
    protected function afterExecute(array $config, array $context, mixed $result): void
    {
        if ($this->debugMode) {
            error_log("Completed @{$this->getName()} with result: " . json_encode($result));
        }
    }
    
    /**
     * Handle execution errors
     */
    protected function handleError(\Exception $e, array $config, array $context): void
    {
        $errorContext = [
            'operator' => $this->getName(),
            'config' => $config,
            'context' => $context,
            'error' => $e->getMessage(),
            'trace' => $e->getTraceAsString()
        ];
        
        error_log("Operator @{$this->getName()} error: " . json_encode($errorContext));
        
        // Emit metrics if available
        $this->emitErrorMetric($e);
    }
    
    /**
     * Emit error metric
     */
    protected function emitErrorMetric(\Exception $e): void
    {
        // Override in subclasses to emit specific error metrics
    }
    
    /**
     * Get context value with fallback
     */
    protected function getContextValue(array $context, string $key, mixed $default = null): mixed
    {
        return $context[$key] ?? $default;
    }
    
    /**
     * Resolve variable references
     */
    protected function resolveVariable(mixed $value, array $context): mixed
    {
        if (is_array($value) && isset($value['variable'])) {
            $varName = $value['variable'];
            return $this->getContextValue($context, $varName);
        }
        
        return $value;
    }
    
    /**
     * Set debug mode
     */
    public function setDebugMode(bool $enabled): void
    {
        $this->debugMode = $enabled;
    }
    
    /**
     * Cleanup resources
     */
    public function cleanup(): void
    {
        // Override in subclasses to cleanup resources
    }
    
    /**
     * Create structured error response
     */
    protected function createErrorResponse(string $code, string $message, array $context = []): array
    {
        return [
            'error' => $code,
            'operator' => "@{$this->getName()}",
            'message' => $message,
            'context' => $context,
            'suggestions' => $this->getErrorSuggestions($code),
            'trace_id' => $this->generateTraceId()
        ];
    }
    
    /**
     * Get error suggestions
     */
    protected function getErrorSuggestions(string $code): array
    {
        $suggestions = [
            'INVALID_CONFIG' => [
                'Check the operator documentation for required fields',
                'Verify all field types are correct',
                'Ensure all required dependencies are available'
            ],
            'EXECUTION_FAILED' => [
                'Check external service availability',
                'Verify authentication credentials',
                'Review error logs for detailed information'
            ],
            'TIMEOUT' => [
                'Increase timeout configuration',
                'Check network connectivity',
                'Consider using retry logic'
            ],
            'CONNECTION_FAILED' => [
                'Verify service endpoint is correct',
                'Check firewall and network settings',
                'Ensure authentication is properly configured'
            ]
        ];
        
        return $suggestions[$code] ?? ['Review the operator documentation for troubleshooting'];
    }
    
    /**
     * Generate trace ID
     */
    protected function generateTraceId(): string
    {
        return uniqid($this->getName() . '-', true);
    }
    
    /**
     * Log operation
     */
    protected function log(string $level, string $message, array $context = []): void
    {
        $logEntry = [
            'timestamp' => date('Y-m-d H:i:s'),
            'level' => $level,
            'operator' => $this->getName(),
            'message' => $message,
            'context' => $context
        ];
        
        error_log(json_encode($logEntry));
    }
} 

File: sdk/csharp/csharp-sdk-package-20250721_013844/operators/Operators/BaseOperator.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Threading.Tasks;

namespace TuskLang.Operators
{
    /// <summary>
    /// Base Operator Class for TuskLang C# SDK
    /// 
    /// Provides common functionality for all @ operators including
    /// validation, error handling, and context management.
    /// 
    /// Follows the same pattern as the PHP SDK BaseOperator.
    /// </summary>
    public abstract class BaseOperator
    {
        protected Dictionary<string, object> DefaultConfig { get; set; } = new Dictionary<string, object>();
        protected List<string> RequiredFields { get; set; } = new List<string>();
        protected List<string> OptionalFields { get; set; } = new List<string>();
        protected string Version { get; set; } = "1.0.0";
        protected bool DebugMode { get; set; } = false;
        
        /// <summary>
        /// Get operator name
        /// </summary>
        public abstract string GetName();
        
        /// <summary>
        /// Get operator version
        /// </summary>
        public virtual string GetVersion() => Version;
        
        /// <summary>
        /// Get operator description
        /// </summary>
        protected virtual string GetDescription() => "Base operator description - override in subclasses";
        
        /// <summary>
        /// Get usage examples
        /// </summary>
        protected virtual Dictionary<string, string> GetExamples()
        {
            return new Dictionary<string, string>
            {
                ["basic"] = $"@{GetName()}(\"example\")",
                ["advanced"] = $"@{GetName()}({{param: \"value\"}})"
            };
        }
        
        /// <summary>
        /// Get error codes
        /// </summary>
        protected virtual Dictionary<string, string> GetErrorCodes()
        {
            return new Dictionary<string, string>
            {
                ["INVALID_CONFIG"] = "Invalid configuration provided",
                ["EXECUTION_FAILED"] = "Operator execution failed",
                ["TIMEOUT"] = "Operation timed out",
                ["CONNECTION_FAILED"] = "Connection to external service failed"
            };
        }
        
        /// <summary>
        /// Get operator schema
        /// </summary>
        public virtual Dictionary<string, object> GetSchema()
        {
            return new Dictionary<string, object>
            {
                ["name"] = GetName(),
                ["version"] = GetVersion(),
                ["description"] = GetDescription(),
                ["required_fields"] = RequiredFields,
                ["optional_fields"] = OptionalFields,
                ["default_config"] = DefaultConfig,
                ["examples"] = GetExamples(),
                ["error_codes"] = GetErrorCodes()
            };
        }
        
        /// <summary>
        /// Validate configuration
        /// </summary>
        public virtual ValidationResult Validate(Dictionary<string, object> config)
        {
            var errors = new List<string>();
            var warnings = new List<string>();
            
            // Check required fields
            foreach (var field in RequiredFields)
            {
                if (!config.ContainsKey(field))
                {
                    errors.Add($"Required field '{field}' is missing");
                }
            }
            
            // Validate field types and values
            var validation = ValidateFields(config);
            errors.AddRange(validation.Errors);
            warnings.AddRange(validation.Warnings);
            
            // Custom validation
            var customValidation = CustomValidate(config);
            errors.AddRange(customValidation.Errors);
            warnings.AddRange(customValidation.Warnings);
            
            return new ValidationResult
            {
                IsValid = errors.Count == 0,
                Errors = errors,
                Warnings = warnings
            };
        }
        
        /// <summary>
        /// Validate individual fields
        /// </summary>
        protected virtual ValidationResult ValidateFields(Dictionary<string, object> config)
        {
            var errors = new List<string>();
            var warnings = new List<string>();
            
            foreach (var kvp in config)
            {
                var validation = ValidateField(kvp.Key, kvp.Value);
                errors.AddRange(validation.Errors);
                warnings.AddRange(validation.Warnings);
            }
            
            return new ValidationResult { Errors = errors, Warnings = warnings };
        }
        
        /// <summary>
        /// Validate a single field
        /// </summary>
        protected virtual ValidationResult ValidateField(string field, object value)
        {
            var errors = new List<string>();
            var warnings = new List<string>();
            
            // Check if field is known
            if (!RequiredFields.Contains(field) && !OptionalFields.Contains(field))
            {
                warnings.Add($"Unknown field '{field}'");
            }
            
            // Type validation
            var typeValidation = ValidateFieldType(field, value);
            errors.AddRange(typeValidation.Errors);
            
            return new ValidationResult { Errors = errors, Warnings = warnings };
        }
        
        /// <summary>
        /// Validate field type
        /// </summary>
        protected virtual ValidationResult ValidateFieldType(string field, object value)
        {
            // Override in subclasses for specific type validation
            return new ValidationResult();
        }
        
        /// <summary>
        /// Custom validation logic
        /// </summary>
        protected virtual ValidationResult CustomValidate(Dictionary<string, object> config)
        {
            // Override in subclasses for custom validation
            return new ValidationResult();
        }
        
        /// <summary>
        /// Execute operator
        /// </summary>
        public virtual async Task<object> ExecuteAsync(Dictionary<string, object> config, Dictionary<string, object> context)
        {
            try
            {
                // Merge with default configuration
                var mergedConfig = new Dictionary<string, object>(DefaultConfig);
                foreach (var kvp in config)
                {
                    mergedConfig[kvp.Key] = kvp.Value;
                }
                
                // Pre-execution hooks
                BeforeExecute(mergedConfig, context);
                
                // Execute the operator
                var result = await ExecuteOperatorAsync(mergedConfig, context);
                
                // Post-execution hooks
                AfterExecute(mergedConfig, context, result);
                
                return result;
            }
            catch (Exception ex)
            {
                HandleError(ex, config, context);
                throw;
            }
        }
        
        /// <summary>
        /// Execute operator (synchronous version)
        /// </summary>
        public virtual object Execute(Dictionary<string, object> config, Dictionary<string, object> context)
        {
            return ExecuteAsync(config, context).GetAwaiter().GetResult();
        }
        
        /// <summary>
        /// Execute the actual operator logic
        /// </summary>
        protected abstract Task<object> ExecuteOperatorAsync(Dictionary<string, object> config, Dictionary<string, object> context);
        
        /// <summary>
        /// Pre-execution hook
        /// </summary>
        protected virtual void BeforeExecute(Dictionary<string, object> config, Dictionary<string, object> context)
        {
            if (DebugMode)
            {
                Log("debug", $"Executing {GetName()} operator", new Dictionary<string, object>
                {
                    ["config"] = config,
                    ["context"] = context
                });
            }
        }
        
        /// <summary>
        /// Post-execution hook
        /// </summary>
        protected virtual void AfterExecute(Dictionary<string, object> config, Dictionary<string, object> context, object result)
        {
            if (DebugMode)
            {
                Log("info", $"{GetName()} operator completed", new Dictionary<string, object>
                {
                    ["result_type"] = result?.GetType().Name
                });
            }
        }
        
        /// <summary>
        /// Handle errors
        /// </summary>
        protected virtual void HandleError(Exception ex, Dictionary<string, object> config, Dictionary<string, object> context)
        {
            Log("error", $"Operator {GetName()} failed", new Dictionary<string, object>
            {
                ["error"] = ex.Message,
                ["stack_trace"] = ex.StackTrace,
                ["config"] = config
            });
            
            EmitErrorMetric(ex);
        }
        
        /// <summary>
        /// Emit error metric
        /// </summary>
        protected virtual void EmitErrorMetric(Exception ex)
        {
            // Override in subclasses to emit metrics
        }
        
        /// <summary>
        /// Get context value
        /// </summary>
        protected virtual T GetContextValue<T>(Dictionary<string, object> context, string key, T defaultValue = default)
        {
            if (context.TryGetValue(key, out var value))
            {
                if (value is T typedValue)
                {
                    return typedValue;
                }
                
                try
                {
                    return (T)Convert.ChangeType(value, typeof(T));
                }
                catch
                {
                    return defaultValue;
                }
            }
            
            return defaultValue;
        }
        
        /// <summary>
        /// Resolve variable references
        /// </summary>
        protected virtual object ResolveVariable(object value, Dictionary<string, object> context)
        {
            if (value is string strValue)
            {
                // Handle @variable references
                if (strValue.StartsWith("@variable(") && strValue.EndsWith(")"))
                {
                    var varName = strValue.Substring(10, strValue.Length - 11).Trim('"', '\'');
                    if (context.TryGetValue(varName, out var varValue))
                    {
                        return varValue;
                    }
                }
                
                // Handle @env references
                if (strValue.StartsWith("@env(") && strValue.EndsWith(")"))
                {
                    var envVar = strValue.Substring(5, strValue.Length - 6).Trim('"', '\'');
                    return Environment.GetEnvironmentVariable(envVar) ?? "";
                }
            }
            
            return value;
        }
        
        /// <summary>
        /// Set debug mode
        /// </summary>
        public virtual void SetDebugMode(bool enabled)
        {
            DebugMode = enabled;
        }
        
        /// <summary>
        /// Cleanup resources
        /// </summary>
        public virtual void Cleanup()
        {
            // Override in subclasses to cleanup resources
        }
        
        /// <summary>
        /// Create error response
        /// </summary>
        protected virtual Dictionary<string, object> CreateErrorResponse(string code, string message, Dictionary<string, object> context = null)
        {
            return new Dictionary<string, object>
            {
                ["error"] = true,
                ["code"] = code,
                ["message"] = message,
                ["timestamp"] = DateTime.UtcNow,
                ["trace_id"] = GenerateTraceId(),
                ["context"] = context ?? new Dictionary<string, object>()
            };
        }
        
        /// <summary>
        /// Get error suggestions
        /// </summary>
        protected virtual List<string> GetErrorSuggestions(string code)
        {
            var suggestions = new List<string>();
            
            switch (code)
            {
                case "INVALID_CONFIG":
                    suggestions.Add("Check that all required fields are provided");
                    suggestions.Add("Verify field types match expected values");
                    suggestions.Add("Review operator documentation for correct usage");
                    break;
                case "EXECUTION_FAILED":
                    suggestions.Add("Check external service connectivity");
                    suggestions.Add("Verify authentication credentials");
                    suggestions.Add("Review error logs for specific details");
                    break;
                case "TIMEOUT":
                    suggestions.Add("Increase timeout value in configuration");
                    suggestions.Add("Check network connectivity");
                    suggestions.Add("Verify external service is responding");
                    break;
                case "CONNECTION_FAILED":
                    suggestions.Add("Verify connection parameters (host, port, etc.)");
                    suggestions.Add("Check firewall and network settings");
                    suggestions.Add("Ensure external service is running");
                    break;
            }
            
            return suggestions;
        }
        
        /// <summary>
        /// Generate trace ID
        /// </summary>
        protected virtual string GenerateTraceId()
        {
            return Guid.NewGuid().ToString("N");
        }
        
        /// <summary>
        /// Log message
        /// </summary>
        protected virtual void Log(string level, string message, Dictionary<string, object> context = null)
        {
            var logEntry = new Dictionary<string, object>
            {
                ["timestamp"] = DateTime.UtcNow,
                ["level"] = level,
                ["operator"] = GetName(),
                ["message"] = message,
                ["trace_id"] = GenerateTraceId()
            };
            
            if (context != null)
            {
                foreach (var kvp in context)
                {
                    logEntry[kvp.Key] = kvp.Value;
                }
            }
            
            // In a real implementation, this would go to a proper logging system
            if (DebugMode || level == "error")
            {
                Console.WriteLine($"[{level.ToUpper()}] {GetName()}: {message}");
            }
        }
    }
    
    /// <summary>
    /// Validation result
    /// </summary>
    public class ValidationResult
    {
        public bool IsValid { get; set; } = true;
        public List<string> Errors { get; set; } = new List<string>();
        public List<string> Warnings { get; set; } = new List<string>();
    }
} 